kdc/kerberos5.c

```diff
@@ -2512,11 +2512,11 @@ _kdc_as_rep(astgs_request_t r)
          */
         if (r->pa_max_life > 0)
             t = rk_time_add(start, min(rk_time_sub(t, start), r->pa_max_life));
-        else if (r->client->max_life && *r->client->max_life)
+        else if (r->client->max_life)
```

Comment:
The reason for this change is to support authentication policies from Windows, which allow administrators to directly configure the maximum TGT lifetime and renew time to any value, including zero and negative values. For zero and negative maximum lifetimes to behave consistently with Windows, and produce the same error codes, we have to be able to encode these values in the HDB entry, and treat them as legitimate lifetime values (rather than behaving as if no maximum lifetime were enforced).

Reply:
What does AD do with negative ticket max life/renew values?  What is the problem with us clamping negative values to zero?  That is, how visible and annoying a difference from AD's handling would that be?

Also, would we not have an HDB migration problem here?  I think at the very least we'd want to do something like change to 64-bit signed `INTEGER` here so that we can remain backwards compatible with existing HDBs where these max values might be set to values `> INT32_MAX`.

Reply:
> What does AD do with negative ticket max life/renew values? What is the problem with us clamping negative values to zero? That is, how visible and annoying a difference from AD's handling would that be?

Setting a negative maximum lifetime in Windows results in a NEVER_VALID error when performing an AS-REQ, whereas Heimdal’s current behaviour is to treat a non-positive lifetime as if no maximum lifetime were set.

Reply:
Would disabling the principal maybe result in the same behavior?  I believe currently it yields a principal unknown error (client or service, as appropriate), but we could have a param to make the kdc return a different error.